sweep/fee_bumper.go

```diff
@@ -0,0 +1,894 @@
+
+	// fee is the fee paid by the tx.
+	fee btcutil.Amount
+}
+
+// Start starts the publisher by subscribing to block epoch updates and kicking
+// off the monitor loop.
+func (t *TxPublisher) Start() error {
+	log.Info("TxPublisher starting...")
+
+	blockEvent, err := t.cfg.Notifier.RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		return fmt.Errorf("register block epoch ntfn: %v", err)
+	}
+
+	select {
+	case bestBlock := <-blockEvent.Epochs:
+		t.currentHeight = bestBlock.Height
+
+	case <-t.quit:
+		log.Debugf("TxPublisher shutting down, exit start process")
+		return nil
+	}
+
+	t.wg.Add(1)
+	go t.monitor(blockEvent)
+
+	return nil
+}
+
+// Stop stops the publisher and waits for the monitor loop to exit.
+func (t *TxPublisher) Stop() {
+	log.Info("TxPublisher stopping...")
+
+	t.wg.Wait()
+	close(t.quit)
+}
+
+// monitor is the main loop driven by new blocks. Whevenr a new block arrives,
+// it will examine all the txns being monitored, and check if any of them needs
+// to be bumped. If so, it will attempt to bump the fee of the tx.
+//
+// NOTE: Must be run as a goroutine.
+func (t *TxPublisher) monitor(blockEvent *chainntnfs.BlockEpochEvent) {
+	defer blockEvent.Cancel()
+	defer t.wg.Done()
+
+	select {
+	case epoch, ok := <-blockEvent.Epochs:
+		if !ok {
+			// We should stop the publisher before stopping the
+			// chain service. Otherwise it indicates an error.
+			log.Error("Block epoch channel closed, exit monitor")
+
+			return
+		}
+
+		// Update the best known height for the publisher.
+		t.currentHeight = epoch.Height
+
+		// Check all monitored txns to see if any of them needs to be
+		// bumped.
+		t.processRecords()
+
+	case <-t.quit:
+		log.Debug("Fee bumper stopped, exit monitor")
+		return
+	}
+}
+
+// processRecords checks all the txns being monitored, and checks if any of
+// them needs to be bumped. If so, it will attempt to bump the fee of the tx.
+func (t *TxPublisher) processRecords() {
+	// confirmedRecords stores a map of the records which have been
+	// confirmed.
+	confirmedRecords := make(map[uint64]*monitorRecord)
+
+	// feeBumpRecords stores a map of the records which need to be bumped.
+	feeBumpRecords := make(map[uint64]*monitorRecord)
+
+	// visitor is a helper closure that visits each record and divides them
+	// into two groups.
+	visitor := func(requestID uint64, r *monitorRecord) error {
+		log.Tracef("Checking monitor recordID=%v for tx=%v", requestID,
+			r.tx.TxHash())
+
+		// If the tx is already confirmed, we can stop monitoring it.
+		if isConfirmed(r.confEvent) {
+			confirmedRecords[requestID] = r
+
+			// Move to the next record.
+			return nil
+		}
+
+		feeBumpRecords[requestID] = r
+
+		// Return nil to move to the next record.
+		return nil
+	}
+
+	// Iterate through all the records and divide them into two groups.
+	t.records.ForEach(visitor)
+
+	// For records that are confirmed, we'll notify the caller about this
+	// result.
+	for requestID, r := range confirmedRecords {
```

Comment:
chatted offline but leaving note: need to do `rec := r` in case `loopvar` is off? here and below

Reply:
fixed.